### ISSUE TITLE
GeflUdpAppender throws ClosedChannelException

### ISSUE THREAD

**Issue by pfeiler** [osiegmar/logback-gelf#20]
Hi, when logging to Graylog with logback-gelf-1.1.0.jar it stops logging after a minute.
By debugging I see that gelf throws a ClosedChannelException. Maybe gelf needs to check the state of channel and reopen if closed?:

```
[Ljava.lang.StackTraceElement;@39e25fb8java.nio.channels.ClosedChannelException
	at sun.nio.ch.DatagramChannelImpl.ensureOpen(DatagramChannelImpl.java:320)
	at sun.nio.ch.DatagramChannelImpl.send(DatagramChannelImpl.java:428)
	at de.siegmar.logbackgelf.GelfUdpAppender.appendMessage(GelfUdpAppender.java:81)
	at de.siegmar.logbackgelf.AbstractGelfAppender.append(AbstractGelfAppender.java:106)
	at de.siegmar.logbackgelf.AbstractGelfAppender.append(AbstractGelfAppender.java:28)
	at ch.qos.logback.core.UnsynchronizedAppenderBase.doAppend(UnsynchronizedAppenderBase.java:84)
	at ch.qos.logback.core.spi.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:51)
	at ch.qos.logback.classic.Logger.appendLoopOnAppenders(Logger.java:270)
	at ch.qos.logback.classic.Logger.callAppenders(Logger.java:257)
	at ch.qos.logback.classic.Logger.buildLoggingEventAndAppend(Logger.java:421)
	at ch.qos.logback.classic.Logger.filterAndLog_2(Logger.java:414)
	at ch.qos.logback.classic.Logger.info(Logger.java:587)
```

**Comment by osiegmar** [comment]
From my understanding this can only happen if the DatagramChannel has been closed by the method `de.siegmar.logbackgelf.GelfUdpAppender#close` as the channel itself is not connected to a remote system. This close method is called by logback when it is stopping (commonly because of shutting down the whole application). Could that be the case in your scenario?

Is the exception thrown all the times after this first minute or does it stop at some point?

Furthermore, the whole stack trace would be interesting. It might contain additional information...

**Comment by pfeiler** [comment]
Thanks for having a look into this.
- Exception is thrown with every log statement after channel has closed. 
- The channel closes always at the same log statement, why I don't know. 
- The next entries in the stacktrace are from classes from our application (can't disclose). 

I googled a bit, would you please have a look at: 
https://github.com/t0xa/gelfj/issues/87#issuecomment-99060182 and https://github.com/t0xa/gelfj/commit/15842b8b7d0950cbf2d67dbe028fe2e4fb328122

**Comment by osiegmar** [comment]
Ok, I need some more time to investigate. I tried to reproduce this by shutting down the graylog server and starting it again, but as expected it did not result in any exception. What operating system and Java version do you use with logback-gelf?

**Comment by osiegmar** [comment]
Cannot reproduce here. Can you try the branch https://github.com/osiegmar/logback-gelf/tree/udp-reopen (version 1.1.1-SNAPSHOT)? I'm testing for `isOpen()` here and reopen the channel if required.

**Comment by etluchs** [comment]
similar problem here. Only it appears after quite long time (weeks) and only sporadic. Will try the "robust" version ...

**Comment by osiegmar** [comment]
Did the udp-reopen branch solved the issue for someone?

**Comment by pfeiler** [comment]
I am sorry, I could not test it. We switched to TCP which works with no
problems so far. Please close and reopen when needed.

Oliver Siegmar <notifications@github.com> schrieb am Fr., 28. Sep. 2018,
20:48:

**Comment by etluchs** [comment]
Similar here :-( sorry. For the installation in question, we switched to logging to stdout and let docker the forwarding to graylog. 
We may still use it for one of the non-docker environments though ...

**Comment by osiegmar** [comment]
Please reopen this ticket if someone can test the branch

**Comment by Rustam-Kadyrov** [comment]
Had similar problem with version 1.x under docker container. After certain activities GELF appender stopped to send messages to Graylog with exception: java.nio.channels.ClosedChannelException.
Unfortunately I have no chance to check your branch on prod.

**Comment by chshawkn** [comment]
> Had similar problem with version 1.x under docker container. After certain activities GELF appender stopped sending messages to Graylog with exception: java.nio.channels.ClosedChannelException.
> Unfortunately I have no chance to check your branch on prod.

+1

**Comment by skreuzhuber** [comment]
We had the very same issue and have tested your branch successfully.

We had lost GELF messages after an uptime of about 1-2 days, e.g. no more GELF messages were sent by our applications. In our log files we have encountered the stated ClosedChannelException. We use the GELF Appender within various Spring Boot applications that are deployed as Docker containers and send logs via GELF to Graylog.

This issue was solved by building and including your fix (https://github.com/osiegmar/logback-gelf/tree/udp-reopen). Since about 2 weeks no GELF messages were lost on various environments.

Thanks a lot for your effort. We would kindly appreciate if you merge the fix into the master and include it within one of your next releases. Thanks!
